### ISSUE TITLE
Work on an 'Economy Adapter'

### ISSUE THREAD

**Issue by lokka30** [lokka30/Treasury#12]
## About
An Economy Adapter is a separate API from Treasury's main economy API. This API will allow a plugin to use one dependency which will automatically use Treasury or Vault on a server, with very similar methods to Vault's. This will make it extremely easy for a plugin to support Treasury, if it is already too difficult to add direct Treasury support.

Of course, using Vault, the Economy Adapter will be forcefully limited to only being able to offer Vault's Economy API's features since Treasury has them all anyways.

This issue was suggested by [Jikoo](https://www.spigotmc.org/members/jikoo.514/) with this post:

> I would probably want a generic utility supporting Treasury, then falling through to Vault/whatever else pre-written for me to consider adopting it, honestly. This might even be your best bet for wider adoption - a really well-written plug-and-play utility would remove a lot of the horribly janky "definitely optional dependency" handling I've seen from people (myself included, shame on past me) adopting Vault.

The Economy Adapter will also do some inherent improvements to Vault, e.g. forcing the use of UUIDs wherever Treasury does.

A crucial change in this API is that non-player accounts (in Vault this is Bank Accounts) must require an owning member as Vault's API requires this.

## Developer thoughts wanted
I'd like to hear what other developers think about this suggestion, and especially if there are any comments on my plan.

## Explanation of choice of labels
* `thoughts wanted`
  * Interested to hear what you all think about this. even if it's just a thumbs up or something :)
* `unassigned target version`
  * no idea when I will work on this.
* `status: on hold`
  * the economy API needs to be finalized before I will start working on this. See #100

**Comment by lokka30** [comment]
Actually, I think this would be better suited to exist in #63.

**Comment by MrNemo64** [comment]
> Actually, I think this would be better suited to exist in #63.

I think we first need to design this.
For example, Treasury supports multy-currency but Vault doesn't, how are we going to solve that? This _bridge_ between Vault and Treasury may need to store its own data depending on how we do things

**Comment by lokka30** [comment]
I think the economy adapter needs to just be limited in features. e.g. Treasury has no multi-world support, Vault has no multi-currency support, Vault requires bank accounts to have an owner.

**Comment by lokka30** [comment]
Here is my current thought on the Economy Adapter (EA).

* it will be a separate repository & resource implementing vault and treasury
* it will have multi currency support - if Vault is being used then only once currency can be loaded of course
* it will have multi world support - if Treasury is being used then it will iterate through each currency to find a specific currency for a certain world (by checking what primary world it uses).
* it will only be available on Bukkit as Vault is Bukkit-only
* it will have two interfaces: one that has methods for concurrency via using CompletableFutures, another which runs everything synchronously like Vault. I know this isn't good but it will make it easier for developers to justify a transition to it especially if they are using vault because they only have to change their vault api calls, not the rest of the entire code
* non-player accounts will require a member due to Vault's bank accounts requiring a member
* no transaction or transaction history support due to Vault lacking it

**Comment by lokka30** [comment]
@MrNemo64 @Jikoo @MrIvanPlays what do you think about the comment above?

@creatorfromhell, do you think it should also work for Reserve?

**Comment by lokka30** [comment]
Ehhh.. after a bigger think I am really unsure about making this.

Was about to post this on Spigot but I'll wait until more thoughts roll in.
```
I am thinking of making another resource: EconomyAdapter. This was suggested by Jikoo a few months ago.

I believe that there will be developers who can't justify updating their current code to also support Treasury due to the concurrency changes, or even just new plugins which don't want to manage multiple economy API dependencies

EconomyAdapter is a separate API from Treasury and Vault. What it does, is route these API calls to Treasury or Vault, depending on which is installed on the server. This means that developers who are either lazy or have projects which are too difficult to update can utilize Treasury and Vault at the same time through a single dependency.

EconomyAdapter will automatically route these calls to Treasury or Vault, if they are installed.

[U]This will not be a replacement for using Treasury[/U], as EconomyAdapter will be limited by the feature set of Vault.

Here are my current thoughts on that API:
[LIST]
[*]it will be a separate repository & resource implementing vault and treasury
[*]it will have multi currency support - if Vault is being used then only once currency can be loaded of course
[*]it will have multi world support - if Treasury is being used then it will iterate through each currency to find a specific currency for a certain world (by checking what primary world it uses).
[*]it will only be available on Bukkit as Vault is Bukkit-only
[*]it will have two interfaces: one that has methods for concurrency via using CompletableFutures, another which runs everything synchronously like Vault. I know this isn't good but it will make it easier for developers to justify a transition to it especially if they are using vault because they only have to change their vault api calls, not the rest of the entire code
[*]non-player accounts will require a member due to Vault's bank accounts requiring a member
[*]no transaction or transaction history support due to Vault lacking it
[/LIST]

Therefore, plugin developers utilizing EconomyAdapter will only need to shade EconomyAdapter into their plugin.
```

**Comment by MrNemo64** [comment]
I think the Economy Adapter should be a plugin itself, not a dependecy that you have to shade. 
When the plugin loads up seaches for Vault, if it finds it id registers itself as the provider so any consumer plugin will detect the Economy Adapter as the provider, internally the Economy Adapter will only reroute the calls to vault.
This way consumer plugins just need to depend on the treasury API not another one.

**Comment by lokka30** [comment]
> I think the Economy Adapter should be a plugin itself, not a dependecy that you have to shade. When the plugin loads up seaches for Vault, if it finds it id registers itself as the provider so any consumer plugin will detect the Economy Adapter as the provider, internally the Economy Adapter will only reroute the calls to vault. This way consumer plugins just need to depend on the treasury API not another one.

It is unnecessary to:
* make it a plugin dependency instead of being shaded
  * since it does not need to register anything, just route EconomyAdapter API calls to treasury or vault :) 
* register as an economy provider
  * since it does not operate on the provider level
  
This is because the economy adapter is only designed for *consumer* plugins to access providers, not the other way around. :+1:

**Comment by MrNemo64** [comment]
> make it a plugin dependency instead of being shaded
> * since it does not need to register anything, just route EconomyAdapter API calls to treasury or vault :)

It wouldn't be a plugin dependency. A consumer would only depend on the Treasury API not this plugin.


> register as an economy provider
> * since it does not operate on the provider level

It works with the provider (Vault) so it's kinda like a pseudo provider, but I can understand this point



> This is because the economy adapter is only designed for _consumer_ plugins to access providers, not the other way around. 👍

It would be designed only for consumer plugins

**Comment by MrIvanPlays** [comment]
In my opinion this should be a single interface with implementations with features limited to what vault offers. That way plugins can utilize treasury if it is present or if treasury's not present then utilize vault. The minus of such adapter is that there might not be much raw treasury api adoption because of that thing.

**Comment by MrNemo64** [comment]
> In my opinion this should be a single interface with implementations with features limited to what vault offers. That way plugins can utilize treasury if it is present or if treasury's not present then utilize vault. The minus of such adapter is that there might not be much raw treasury api adoption because of that thing.

But wouldn't consumer plugins have to check for vault?

**Comment by MrIvanPlays** [comment]
> > In my opinion this should be a single interface with implementations with features limited to what vault offers. That way plugins can utilize treasury if it is present or if treasury's not present then utilize vault. The minus of such adapter is that there might not be much raw treasury api adoption because of that thing.
> 
> But wouldn't consumer plugins have to check for vault?

Nope. The way the interface instance is taken will do.

e.g.

```java
public interface EconomyAdapter {
    // ...
}

public final class EconomyAdapterProvider {

    private static EconomyAdapterProvider instance;
    
    public static EconomyAdapterProvider get() {
        if (instance == null) {
            if (Bukkit.getPluginManager().isPluginEnabled("Vault")) {
                instance = new EconomyAdapterVault();
            } else if (Bukkit.getPluginManager().isPluginEnabled("Treasury")) {
                instance = new EconomyAdapterTreasury();
            } else {
                instance = new NoopEconomyAdapter();
            }
        }
        return instance;
    }

}
```

**Comment by lokka30** [comment]
> > > In my opinion this should be a single interface with implementations with features limited to what vault offers. That way plugins can utilize treasury if it is present or if treasury's not present then utilize vault. The minus of such adapter is that there might not be much raw treasury api adoption because of that thing.
> > 
> > 
> > But wouldn't consumer plugins have to check for vault?
> 
> Nope. The way the interface instance is taken will do.
> 
> e.g.
> 
> ```java
> public interface EconomyAdapter {
>     // ...
> }
> 
> public final class EconomyAdapterProvider {
> 
>     private static EconomyAdapterProvider instance;
>     
>     public static EconomyAdapterProvider get() {
>         if (instance == null) {
>             if (Bukkit.getPluginManager().isPluginEnabled("Vault")) {
>                 instance = new EconomyAdapterVault();
>             } else if (Bukkit.getPluginManager().isPluginEnabled("Treasury")) {
>                 instance = new EconomyAdapterTreasury();
>             } else {
>                 instance = new NoopEconomyAdapter();
>             }
>         }
>         return instance;
>     }
> 
> }
> ```

Exactly this. No hooking is needed or pseudo-providers :) this all operates at the consumer level, so all the plugin needs to do is contact the service manager for Bukkit and Treasury (Treasury has its own, yes) and check which one is installed.

**Comment by lokka30** [comment]
I also agree with having a single interface, just like Vault. I want to make it as easy as possible for developers to be able to utilize Treasury, if they are already uncomfortable with directly switching to it.

**Comment by lokka30** [comment]
This could be included inside the Treasury project as a separate module. Thoughts?

**Comment by lokka30** [comment]
@MrIvanPlays @MrNemo64 @Jikoo 
What do you all think about having the Economy Adapter as a separate module within Treasury?

(This is on hold until #100 is completed)

**Comment by MrNemo64** [comment]
Seeing that this will be and interface and not much more being a module is probably the best option

**Comment by lokka30** [comment]
> Seeing that this will be and interface and not much more being a module is probably the best option

My current idea is that it is a class. A plugin instantiates it and then it will check whether Treasury or Vault is installed. If Treasury is installed then that will be prioritized. The class contains methods akin to Vault's which just use the appropriate equivalent in either API.

**Comment by MrNemo64** [comment]
> > Seeing that this will be and interface and not much more being a module is probably the best option
> 
> My current idea is that it is a class. A plugin instantiates it and then it will check whether Treasury or Vault is installed. If Treasury is installed then that will be prioritized. The class contains methods akin to Vault's which just use the appropriate equivalent in either API.
> 
> Otherwise, who would provide the interface?

It would be shaded into the consumer plugin I suppose.
My main issue with this aproach is limitting Treasurys features.

**Comment by lokka30** [comment]
> > > Seeing that this will be and interface and not much more being a module is probably the best option
> > 
> > 
> > My current idea is that it is a class. A plugin instantiates it and then it will check whether Treasury or Vault is installed. If Treasury is installed then that will be prioritized. The class contains methods akin to Vault's which just use the appropriate equivalent in either API.
> > Otherwise, who would provide the interface?
> 
> It would be shaded into the consumer plugin I suppose. My main issue with this aproach is limitting Treasurys features.

I think it's a double edged sword.

Some plugins are so deeply integrated into using Vault that it will be difficult for them to switch to Treasury.

But this API might make lazy developers use it instead of adding an actual integration to Treasury.

I'm fine with not going forward with this although there are definitely use cases with it that are justified.

**Comment by MrNemo64** [comment]
This API should allow you to go from using Vault to the API itself easily but it also needs to make it so going from the API to Treasury is also easy. It has to be a middle step. Thinking of it like that it makes sense to use vault features but with treasurys format, I'm okay with that

**Comment by lokka30** [comment]
> This API should allow you to go from using Vault to the API itself easily but it also needs to make it so going from the API to Treasury is also easy. It has to be a middle step. Thinking of it like that it makes sense to use vault features but with treasurys format, I'm okay with that

What might be better, is if the economy adapter itself is just an interface with methods that are Vault and Treasury compatible. We can also supply a class like `DefaultEconomyAdapter` which implements this interface. That way if people find the default economy adapter suitable then they can use it, otherwise they can create their own economy adapter.
What do you think about that idea?

**Comment by lokka30** [comment]
We have decided not to work on the economy adapter in its entirety.
